### ISSUE TITLE
Rename "id" to "$id" (v5 proposal)

### ISSUE THREAD

**Issue by sam-at-github** [json-schema/json-schema#215]
We already have "$schema", and "$ref". Seems strange that `id` is not "$id". It is semantically part of this trio of meta keywords so should be named similarly. It also make it easier to distinguish between proper id meta keyword and a property named `id`. Even if it is technically possible to distinguish between the two uses of `id`, renaming "id" "$id" would make parsing easier, and also reduce the cognitive load ("now, is that an id or and id..."). Refs https://github.com/json-schema/json-schema/issues/77.

**Comment by Fannon** [comment]
I agree. I was expecting it to be $id when I was learning JSON Schema.

**Comment by robbie-mac** [comment]
+1, that messed me up the first few times as well. It just seemed natural to type $id

**Comment by Acubed** [comment]
Is there a previous established pattern you may be thinking of, or is it strictly because `$schema` is also so named?

Also, note that `$ref` isn't a part of JSON Schema as such. Can we think of any negative consequences of this?

**Comment by robbie-mac** [comment]
It was simply muscle memory. Typed $schema and then typed $id.
After the first couple times I relized I made this error, I correcyrd my behavior.
It just seemed natural that 2 fields so closely tied together would use the same naming convention.
However, now that you mention it, since all three fields ($schema, id, and ref) all point to a resource shouldnt they all share the same naming convention?

I can't think of a negative, but I can think of a posative. By naming all 3 with $ prefix, that would give some separation from fields within the schema itself that may share the same name. So long as the schema author doesnt goof up and name theirs with a $ prefix.

**Comment by YurySk** [comment]
Renaming "id" will break backward compatibility. You are looking at a lot of unhappy people there...

**Comment by JanTvrdik** [comment]
@YurySk Not really. The libraries will not stop supporting draft 4 anytime soon.

**Comment by YurySk** [comment]
So V4 schemas will not be valid in V5? Not sure I like having to convert things like this.

**Comment by JanTvrdik** [comment]
Draft 3 schemas are not valid draft 4 schemas. Keeping the draft 5 fully compatible with draft 4 is impossible anyway due to draft 4 ambiguities.

**Comment by erosb** [comment]
I'm not sure of draft v5 should be compatible with v4. Since the spec is
still a draft, there can be any changes. Draft means instability (at least
for me).

2016-04-22 1:34 GMT+02:00 Jan Tvrdík <notifications@github.com>:

> Draft 3 schemas are not valid draft 4 schemas. Keeping the draft 5 fully
> compatible with draft 4 is impossible anyway due to draft 4 ambiguities.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/json-schema/json-schema/issues/215#issuecomment-213159559>
>



-- 


*Bence Erősme@github <https://github.com/erosb>*

**Comment by robbie-mac** [comment]
Firstly, I have to apologise for all the typos in my last post. Typing on a mobile, whike getting bounced around on public transit does not equal a quality message. Sigh.
@JanTvrdik and @erosb I agree that breaking changes are inevitable.
@YurySk I hear you with backwards compatability, however I find that when that argument is brought up it is not the real issue. If backwards compatability was the top of the requirements list, we might as well close up shop and call it quits. V4 is here and it is 100% Guaranteed to be compatable with V4... except V4 doesnt do everything we want/need. Or a spec article can be interpreted multiple ways, with far reaching consequences.
I think the real issue when backwards compatability is brought up is adoption rate. If changes are to far sweeping people wont want to re-learn. However renaming $schema, or $id and $ref, is not a major change in the grand scheme of things. Add in the fact that V4 can still be used after V5 comes out. Just like V3 and V2 can still be used.

**Comment by YurySk** [comment]
@robbie-mac

I hear you. V4 indeed leaves a lot to be desired (as I learned painfully during my validator implementation). If it becomes a cleaner standard that I could live with having to port schemas to V5.

**Comment by coding2012** [comment]
DV4 is a "defacto standard". Adoption rates of Json schema validation has skyrocketed in the last 8 months. Please write the standard to support id and $id.

http://npm-stat.com/charts.html?package=tv4&author=&from=2015-05-17&to=2016-05-16
http://npm-stat.com/charts.html?package=ajv&author=&from=2015-05-14&to=2016-05-13
http://npm-stat.com/charts.html?package=jsonschema&author=&from=2015-05-13&to=2016-05-12

**Comment by Julian** [comment]
Backwards compatibility is different from API churn :)

There are lots of changes when moving from Draft 3 to Draft 4 that were very well meaning but caused lots of API churn. This one too seems well-motivated, but I'd also caution against it personally. I'd like to be very very careful about which things we change from Draft4 that aren't "broken" specifically.

**Comment by robbie-mac** [comment]
Is tv4 a 'defacto standard' or is chai a popular assertion library?

I would propose that the downloads of tv4 are because of chai's popularity. Not its own merit.

Codifying a standard because of 1 implementation is a little irresponsable if you ask me.


Rob






On Tue, May 17, 2016 at 10:50 AM -0700, "Julian Berman" <notifications@github.com> wrote:












Backwards compatibility is different from API churn :)



There are lots of changes when moving from Draft 3 to Draft 4 that were very well meaning but caused lots of API churn. This one too seems well-motivated, but I'd also caution against it personally. I'd like to be very very careful about which things we change from Draft4 that aren't "broken" specifically.

**Comment by YurySk** [comment]
There is definitely more than one V4 implementation. For java there's json-schema-validator and I also about finished with a closed source version of a V4 validator.

**Comment by robbie-mac** [comment]
And do they also implement both $id and id?


Heres the reason I am against putting both in the spec.

Keyword exhaustion. If we put both in the spec, there wont be a standard version used. Some will use one flavor, while others, well, the other. It will result in confusion while reading some implementations.

Not to mention the spec will have failed at its ultimate purpose which is to codify a standard.

Plus, further confusion will result if people need an ID field in their schema. Can people use one or the other? Sure. Can it cause confusion, absolutely.

There is already confusion around using the meta field names and their purposes.


Now my coment about one inplementation. That was in response to a previous post (my apologies for not quoting or referencing the post) that one implementation has reached defacto standard status based simply on the number of downloads on npm. If we dig a little bit deeper you can find that it might be because it has been made a dependency of a popular assertion library.

If you said to me that all or even a majority of the validation libraries out there had inplemented the same way, ok, thats a defacto standard.


I think its irresponsable to write a spec based on the way one implementation does things. And to say that the spec has to evolve in a particular direction because of it, is contrary to the process.

Concensus drives standards.

Hijacking a spec because 1 person/team wrote a library based on a draft of a spec, but introducing their own ideas. Thats not consensus, thats lock-in.


Thats almost as bad as saying that the spec has to maintain backwards compatability. Heres an example of a spec that has had to maintain backwards compatability. HTML. What a mess that is. I know the reason for it. We dont want the web to break. The reason that the web woukd break if each new versikn didnt maintain backwards compatability? Because it was the defacto standard to NOT include the html version in your page. Even though the spec defined a method of doing so.

JSON-Schema also provides a method to version. Its right there in the $schema field.

Backwards compatability taken care of. Lets move on with consensus.


Rob






On Tue, May 17, 2016 at 11:38 AM -0700, "Yury Skanavy" <notifications@github.com> wrote:












There is definitely more than one V4 implementation. For java there's json-schema-validator and I also about finished with a closed source version of a V4 validator.

**Comment by YurySk** [comment]
> And do they also implement both $id and id?

Mine does not implement $id...I was going by V4 specs so why would it?

**Comment by robbie-mac** [comment]
If it followed the draft spec it wouldnt.


And wheb the next draft changes things, it would follow the next draft.


Im not arguing against keeping it the same or even changing it at this point. The points I am trying to make are these:

1) stating the spec should move in a particular direction based on a limited and most likely skewed sampling pool off of NPM is wrong. Write the spec for technical reasons. (To go back to HTML we got the marqee tag because of defacto standard, we got rid of it through consensus)

2) having both is not achieving consensus, thats avoidance. There should be a clear technical reason for any change.


If I were to argue one way or another, I would support changing it. Have all of the schema meta fields prefixed with $, making $schema, $id...

That gives us some code consistance, and clarity when reading a schema.


Rob






On Tue, May 17, 2016 at 12:29 PM -0700, "Yury Skanavy" <notifications@github.com> wrote:













And do they also implement both $id and id?




Mine does not implement $id...I was going by V4 specs so why would it?

**Comment by YurySk** [comment]
It seems to me that the issue of one char difference in one keyword name should not warrant any kind of turmoil unless someone wants to use it to argue a point or approach.
If the next draft is not backwards compatible than it can go either way, either purist one or practical one that considers the added migration effort incurred by one char. Depends on the how much difference there is overall.

**Comment by robbie-mac** [comment]
Thats quite a divisive, and counter productive comment.


In one comment you've managed to state that the two things are mutually exclusive, "Either purist or practical"


and signalled that people who have a differing opinion are simply bieng argumentative for the sake of turmoil. "one char difference in one keyword name should not warrant any kind of turmoil unless someone wants to use it to argue a point or approach"


So spelling and other typos should never get fixed?


Not only that, you have missed the real crux of the points I have made.


Rob

**Comment by YurySk** [comment]
> So spelling and other typos should never get fixed?
You are arguing a point or approach aren't you? :) 

I did not mean that people who argue about one character do it just for the sake of turmoil. Nothing wrong with using a particular case to crystallize an approach. As long as we are honest about it. So perhaps you are reading too much into my words.

Nothing particularly wrong with consistency and purism (it's not a bad word actually) as long as the impact of a change is considered in terms of amount of work it adds to users. I.e. if there's legacy, it should be considered to a point. If something is ill-defiled (such as $ref stuff) than it obviously needs to be to be fixed even at the expense of a change that might cascade from that. But I am not sure I feel as strongly about a name change. Again, if the amount of changes in V5 is so great that migrating "id" to "$id" drowns in the other stuff that users needs to do than sure why not make it pretty (again, not a bad word). But if a continuity is mostly preserved (perhaps without 100% backwards compatibility) than...well...maybe not.

I really have nothing else to say here. I wasn't even arguing here so I am not sure what's up with all the emotional rigor.

**Comment by Acubed** [comment]
I should comment that development for the next release is happening over at <https://github.com/json-schema-org/json-schema-spec>.

It's my intention that the next release will be only for bugfixes, updates, and necessary fixes of known problems; any new functionality will be withheld for the next release (which would follow up very soon).

**Comment by robbie-mac** [comment]
@YurySk 
Firstly, I have to apologize, again. I have learned a few things, don't respond to github emails by hitting reply, go to the site ( It makes for really crappy looking posts and are hard to follow with the email header crap ) and don't respond while at the airport waiting for a plane. Short temper, and distractions makes for poor judgement in reading posts, and composing replies..
So that said, my question about fixing typos. That was aimed at the comment of fixing single characters, or changing single characters in a spec.
As for the rest, its really irrelevant as I was simply responding to the post about v4 being a defacto standard. How can a draft of a standard, in the process of being refined, be a standard?

The rest/examples were intended as hypothetical proofs.

**Comment by YurySk** [comment]
@robbie-mac that's ok, I understand.
I am not sure what else to say. Formally you are right and 'typos' should get fixed. But if many people need to do many changes to many schemas just for this one character....Which may not even be the case given that there might be other incompatibilities. That's why I personally can't form a clear opinion without seeing the overall picture of what migration would entail.
If most V4 schemas will be valid in V5 than I suggest thinking hard if it's worth fixing this typo. But if this is not the case than it's ok I suppose. **If** you need to change other things you might as well change 'id' to '$id'

**Comment by handrews** [comment]
@sam-at-github :  since this discussion has moved to json-schema-org/json-schema-spec#20 , could you please close this?  There is no longer anyone with the project who has access to close other people's issues on this repo.

**Comment by Relequestual** [comment]
If anyone is still interested in this issue, would be great if you could pop over and give a thumbs up for https://github.com/json-schema-org/json-schema-spec/pull/154 (which looks to fix issue #20 as mentioned in the previous comment here).

@coding2012 made a good point, that it should be backwards compatible initially, which @handrews has made sure of with his PR.

**Comment by Relequestual** [comment]
@sam-at-github It would be great to get your approval on the new PR also! Would be nice to close of issues on this defunct org's repo.
